### PR TITLE
update linter setup and use golangci-lint

### PR DESCRIPTION
**Pull request by gregwebs**
This is a starter for #1152

This is a setup I use in another project and it is quite similar to the setup in pingcap/tidb.

This has the only 2 linter tools we should need, but additional tools used by tidb-operator can be added to this system.

Currently golangci-lint fails.

- [ ] remove linters from the top-level Makefile
- [ ] get golangci-lint to pass (its is possible to tell it to ignore issues for now)

### CODE REVIEW

**Comment by cofyc** [comment]
I don't think we need to compile tools if they have binary releases: https://github.com/golangci/golangci-lint/releases/tag/v1.21.0. We can download it by using curl or wget, e.g. https://github.com/pingcap/tidb-operator/blob/d4aed2a335c1d9617ed5d1d4afc33fbf7fedc0b8/hack/lib.sh#L99-L109.

File: hack/check [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	if ! command -v gfind ; then
+		echo "please install GNU Find"
+		echo ""
+		echo "    brew install findutils"
+		exit 1
+	fi
+	find=gfind
+fi
+
+pushd "$(dirname "$0")/.." >/dev/null
```
Comment (onlymellb) [inline hack/check]: Can you use this [style](https://github.com/pingcap/tidb-operator/blob/master/hack/run-e2e.sh#L7-L8) uniformly?

File: hack/check [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+NO_LINT=${NO_LINT:-""}
+if [[ -z $NO_LINT ]] ; then
+  echo "linting"
+  ./tools/bin/revive -formatter friendly -config revive.toml ${PKGS}
+fi
+
+echo "checking"
+pushd pkg
+for d in * ; do
+  pushd "$d"
```
Comment (onlymellb) [inline hack/check]: Can we use the absolute path instead of switch dir?

**Comment by gregwebs** [comment]
Feel free to take this PR over. What is here is the easy part. More work is to fix all the linter issues.

File: hack/check [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+if [ "$(uname)" == "Darwin" ]; then
+	if ! command -v gfind ; then
+		echo "please install GNU Find"
+		echo ""
+		echo "    brew install findutils"
+		exit 1
+	fi
+	find=gfind
+fi
```
Comment (onlymellb) [inline hack/check]: I think we can omit this code and use `PKGS=$(find pkg -type f -name '*.go'|awk '{print "github.com/pingcap/tidb-operator/"$1}'|sort -u)` instead.

**Comment by gregwebs** [comment]
I agree that using binary releases would work well. I find myself sometimes needing a tool or version of a tool without such a release. Building also comes with security and reproducibility benefits that are probably not a big concern for codegen and linting tools.

**Pull request closed without merging**